// fireblocks-sdk/dist/src/staking/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegationSummaryDtoByVault = exports.StakingChain = exports.PositionStatus = void 0;
var PositionStatus;
(function (PositionStatus) {
    PositionStatus["error"] = "error";
    PositionStatus["creating"] = "creating";
    PositionStatus["failed"] = "failed";
    PositionStatus["pending"] = "pending";
    PositionStatus["canceled"] = "canceled";
    PositionStatus["activating"] = "activating";
    PositionStatus["active"] = "active";
    PositionStatus["deactivating"] = "deactivating";
    PositionStatus["deactivated"] = "deactivated";
    PositionStatus["withdrawing"] = "withdrawing";
    PositionStatus["withdrawn"] = "withdrawn";
})(PositionStatus || (exports.PositionStatus = PositionStatus = {}));
/**
 * Staking actions
 */
/**
 * Staking chain descriptors
 */
var StakingChain;
(function (StakingChain) {
    StakingChain["SOLANA"] = "SOL";
    StakingChain["SOLANA_TESTNET"] = "SOL_TEST";
    StakingChain["ETHEREUM"] = "ETH";
    StakingChain["GOERLI"] = "ETH_TEST3";
    StakingChain["MATIC"] = "MATIC";
})(StakingChain || (exports.StakingChain = StakingChain = {}));
class DelegationSummaryDtoByVault {
}
exports.DelegationSummaryDtoByVault = DelegationSummaryDtoByVault;
